### ISSUE TITLE
Test library against NPM dataset

### ISSUE THREAD

**Issue by zkat** [felipesere/node-semver-rs#4]
Now that everything seems to be in order, it's time to see if the library actually successfully parses the sort of garbage people put out there.

**Comment by skyfallwastaken** [comment]
Hi! I'm willing to try this! Do you have a link to the dataset?

**Comment by zkat** [comment]
Nope. That's really a big part of the challenge. One way to go about this is to follow this tutorial: https://github.com/npm/registry-follower-tutorial, and use https://github.com/npm/concurrent-couch-follower to write a script that can iterate over these, resume if it crashes, etc. There's a **lot** of NPM packages, after all, so you might assume this'll take a while.

**Comment by skyfallwastaken** [comment]
Considering that there are 2M+ packages on npm, I _really_ don't think getting all of them is a good idea... perhaps 2,000 would work?

**Comment by skyfallwastaken** [comment]
Okay, I've got an idea:

What if we used the
```
GET https://registry.npmjs.org/-/v1/search?text=word
```
endpoint? Basically you'd get an English words list of ~50 words, and _slowly_ (perhaps once an hour) replace `word` with an English word, and hit the endpoint. Record the word, and get the versions. Then compare `node_semver`'s parsing to [the semver package that npm uses](https://github.com/npm/node-semver), and somehow raise an alert if there's a difference.

**Comment by zkat** [comment]
honestly, 2M+ sounds scarier than it sounds :).

"All" you need to do for the dataset is to grab the "versions" field from the stream, then go over each version, and collect the values for all of `dependencies`, `devDependencies`, and `peerDependencies` and shove them somewhere. Even a flat file would work. It might take a few hours to a day to stream them all out, I guess, but that's just waiting time. The stream goes by pretty fast. 

My colleagues used to occasionally collect the entire registry index this way. I guess you could stand up a CouchDB of your own and have it replicate NPM. But really, this part of the registry is much smaller than the larger registry, since it's only metadata, not package contents.

**Comment by zkat** [comment]
keep in mind that this bug is precisely about finding the rare, quirky expressions out there, to make sure `node-semver-rs` works on all current NPM packages as-is, so sampling isn't really good enough.

**Comment by skyfallwastaken** [comment]
> "All" you need to do for the dataset is to grab the "versions" field from the stream, then go over each version, and collect the values for all of dependencies, devDependencies, and peerDependencies and shove them somewhere. Even a flat file would work. It might take a few hours to a day to stream them all out, I guess, but that's just waiting time. The stream goes by pretty fast.

Ah, well, that makes the task a lot easier!

I'm not really sure how CouchDB streams work, though. Wouldn't it only give us the version if the package updates? Plenty of packages are published and then never updated.

> keep in mind that this bug is precisely about finding the rare, quirky expressions out there, to make sure node-semver-rs works on all current NPM packages as-is, so sampling isn't really good enough.

Yeah, that makes sense.

**Comment by skyfallwastaken** [comment]
Okay, just implementing downloading all the package names in the registry in this repo: https://github.com/SkyfallWasTaken/node_semver_testing

I ended up ditching the streams idea and just used `https://skimdb.npmjs.com/registry/_all_docs` to get the package names. Streams are probably better, but I don't have enough experience with CouchDB to use them.

Currently, `write-dump` only writes the package names, but `test-dump` should be able to get the versions from the registry easily enough

**Comment by zkat** [comment]
grabbing _all_docs should work too I guess? :)

**Comment by skyfallwastaken** [comment]
Okay, so I've run into a bit of a roadblock and was wondering if you can help:

Right now, I've got everything up to getting the package names done. The issue is that the endpoint I'm using to get the versions of the packages is _slow_ (about 1.5s response time)

Currently, `get-versions-and-ranges` can process about 100,000 packages a day. However, there are _2,147,973_ packages in NPM, so I'd need to keep the PC on for about 2 weeks, and that's not really something I'm looking forward to.

Could you please go over the code and check if there are any more ways to optimize?

https://github.com/SkyfallWasTaken/node_semver_testing/blob/main/src/get_versions_and_ranges.rs

**Comment by zkat** [comment]
you're doing them one at a time. You can probably do a least 20 of them concurrently, which should speed things up significantly.
